returnn/util/sig_proc.py

```diff
@@ -0,0 +1,159 @@
+  ICASSP 2007 and also
+  https://en.wikipedia.org/wiki/Greenwood_function
+  The default values are taken from RASR, see https://github.com/rwth-i6/rasr/blob/master/src/Signal/GammaTone.cc
+  They correspond to the recommended values for human data according to
+  Greenwood, Donald D. "A cochlear frequency‐position function for several species—29 years later."
+  The Journal of the Acoustical Society of America, 1990
+
+  :param float x: fractional length
+  :param float scaling_constant: A in [1]
+  :param float constant_of_integration: k in [1]
+  :param float slope: a in [1]
+  :return: frequency corresponding to given fractional length
+  :rtype: float
+  """
+  assert 0 <= x <= 1, "fractional length has to be between 0 and 1"
+  return scaling_constant * (numpy.power(10.0, slope * x) - constant_of_integration)
+
+
+def inv_greenwood_function(freq, scaling_constant=165.4, constant_of_integration=0.88, slope=2.1):
+  """
+  Inverse greenwood function, convert frequency to fractional length, see
+  Schlueter, Ralf, et al. "Gammatone features and feature combination for large vocabulary speech recognition."
+  ICASSP 2007 and also
+  https://en.wikipedia.org/wiki/Greenwood_function
+  The default values are taken from RASR, see https://github.com/rwth-i6/rasr/blob/master/src/Signal/GammaTone.cc
+  They correspond to the recommended values for human data according to
+  Greenwood, Donald D. "A cochlear frequency‐position function for several species—29 years later."
+  The Journal of the Acoustical Society of America, 1990
+
+  :param float freq: frequency
+  :param float scaling_constant: A in [1]
+  :param float constant_of_integration: k in [1]
+  :param float slope: a in [1]
+  :return: fractional length corresponding to given frequency
+  :rtype: float
+  """
+  return numpy.log10(freq / scaling_constant + constant_of_integration) / slope
+
+
+class GammatoneFilterbank(object):
+  """
+  Class representing a gammatone filterbank.
+  Based on
+  [1] Schlueter, Ralf, et al. "Gammatone features and feature combination for large vocabulary speech recognition."
+  ICASSP 2007
+  """
+
+  def __init__(self, num_channels, length, sample_rate=16000, freq_max=7500., freq_min=100., normalization=True):
+    """
+    :param int num_channels: number of filters
+    :param int length: length of FIR filters in seconds
+    :param int sample_rate: sample rate of audio signal in Hz
+    :param float freq_max: maximum frequency of filterbank
+    :param float freq_min: minimum frequency of filterbank
+    :param bool normalization: normalize filterbanks to maximum frequency response of 0 dB
+    """
+    self.num_channels = num_channels
+    self.length = length
+    self.sample_rate = sample_rate
+    self.freq_max = freq_max
+    self.freq_min = freq_min
+    self.normalization = normalization
+
+  def get_gammatone_filterbank(self):
+    """
+    Returns an array with the parameters of the gammatone filterbank
+
+    :return: gammatone filterbank
```

Comment:
What will be the shape? Is it dependent on `num_channels` and `length`?

Reply:
Yes, it's of shape `(self.length * self.sample_rate, self.num_channels)`. I'll add a comment with the shape